### DISCUSSION TITLE
What's the correct way to use RequestFactory to test the views?

### DISCUSSION THREAD

**Discussion by dvf** [vitalik/django-ninja#1213] [Q&A]
Any help would be lovely.

**Comment by vitalik** [comment]
@dvf could you elaborate more on what you trying to do ?

**Reply by dvf** [reply]
Thanks for the quick response @vitalik (and thanks for the wonderful library!)

I like testing views outside of the middleware because the tests are super fast and very clean. Auth is  simple too. Typically this is done with Django's `RequestFactory`. 

I did a short test with a Ninja paginated view and immediately the `@paginated` decorator generated errors in the request.

So the clearer questions are:

1. Does Django Ninja support testing with `RequestFactory`?
2. If so, what are some best practices for it (I'll commit a PR to the docs)?

**Reply by vitalik** [reply]
@dvf 

Django Ninja has built-in test client that skips middlewares and focuses only on view functionality for faster tests (like full suit of django-ninja 100% coverage takes 2seconds on M1 mac)

you can check more details here - https://django-ninja.dev/guides/testing/

**Reply by dvf** [reply]
@vitalik would appreciate your thoughts and best-practices here. We are using `pytest`.

The project `urls.py` looks like this:

```python
api = NinjaAPI(
    title="My API",
    version=settings.RELEASE_VERSION,
    auth=JWTAuth(),
    docs=Swagger(settings={"persistAuthorization": True}),
)

api.add_router("auth/", auth_router)
api.add_router("accounts/", accounts_router)


@api.exception_handler(APIError)
def api_error(request: HttpRequest, exc: APIError):
    return JsonResponse(
        {"error_code": exc.error_code},
        status=exc.http_status_code,
    )

urlpatterns = [
    path("v2/", api.urls),
]
```

Our `api_client` fixture in `/conftest.py` is:

```python
import pytest
from ninja.testing import TestClient

from project.urls import api


@pytest.fixture
def api_test_client():
    return TestClient(api)
```

An example test looks like:

```python
def test_upsert_nutrients_template_with_id(api_test_client):
    res = api_test_client.get("/foo/test/")
    assert res.status_code == 200
```

Using this fixture in any test results in the following error:

```
E               ninja.errors.ConfigError: Looks like you created multiple NinjaAPIs or TestClients
E               To let ninja distinguish them you need to set either unique version or urls_namespace
E                - NinjaAPI(..., version='2.0.0')
E                - NinjaAPI(..., urls_namespace='otherapi')
E               Already registered: ['api-?.?.?']
```

**Reply by shapiromatron** [reply]
@dvf - I'm having the same issue - did you ever resolve this? I also tried setting up the tests to use the individual routers, not the api, and had a different issue - I kept getting config errors that the Router had already been registered.

**Reply by dvf** [reply]
@shapiromatron I ended up doing this. And not using `RequestFactory`. It's annoying because I like how lightweight request factory is (not to mention you don't need to worry about full auth).

In my root `conftest.py`:
```python
import pytest

from ninja.testing import TestClient
from project.urls import api


@pytest.fixture
def root_ninja_api():
    return api


@pytest.fixture
def root_api_client(root_ninja_api) -> TestClient:
    return TestClient(root_ninja_api)
```

Then I use the fixture in a test:

```python
def test_foo(user, org, root_api_client, mocker):
    res = root_api_client.post(
        "/auth/login/",
        json={
            "username": user.email,
            "password": "1234",
        },
    )
    ...
```
